### PR TITLE
perf: Implement fast paths in critical areas

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [ ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### AI acknowledgment

- [ ] I did _not_ use AI to generate this PR.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Internal performance optimizations to the linting pipeline and formatter hot paths. No public API or behavior changes are intended.

**Measurements.** Numbers below come from a single `hyperfine` invocation per scenario that **interleaves** the baseline (`main` @ 585ef37) and this branch, so machine drift cancels out rather than being attributed to the change. Node v22.19.0, Windows 11.

| Scenario | Baseline | This branch | Speedup |
| --- | --- | --- | --- |
| Loading | 189.6 ms ± 4.2 | 149.5 ms ± 3.8 | **1.27×** (−21.1%) |
| Single File (jshint.js, ~52k reported messages) | 2.163 s ± 0.042 | 1.930 s ± 0.060 | **1.12×** (−10.8%) |
| Multi Files (450 files) | 6.186 s ± 0.398 | 4.829 s ± 0.128 | **1.28×** (−21.9%) |
| **Aggregate** | 8.54 s | 6.91 s | **−19.1%** |

Changes, roughly by impact:

- **`lib/linter/esquery.js` / `lib/linter/source-code-traverser.js`** — selectors consisting only of node-type identifiers/wildcards (the majority registered by core rules) are flagged `alwaysMatches` once bucketed by node type, skipping the `esquery.matches()` call entirely. Adds a specialized fast matcher for the common `Parent > .field` / `Parent > *.field` shape, verified equivalent to `esquery.matches()` (including its case-insensitive identifier comparison) across the selector shapes core rules actually register. Traversal reuses a scratch array instead of allocating a selector array per node and avoids spreading arguments for the common single-argument call.

- **`lib/linter/linter.js` / `lib/linter/source-code-visitor.js`** — removes the per-listener error-handling wrapper closure (a rest-args allocation on the hottest call in the program) when neither `--stats` nor rule timing is enabled. Listeners are tagged with a `RULE_ID` symbol and `SourceCodeVisitor#callSync` attributes thrown errors at the call site. Error attribution (`err.ruleId`) is unchanged, including when two rules share a listener function object and when a thrown error already carries a `ruleId` — both covered by new tests.

- **`lib/rule-tester/rule-tester.js` + `lib/config/config.js`** — `ajv` is constructed lazily in both places. Both are required to matter: `lib/api.js` eagerly requires `RuleTester`, and `api.js → linter → config` is a second independent path to `ajv`. Together they cut the `require("eslint")` graph from **177 modules to 131**, which is what produces the Loading improvement.

- **`lib/linter/file-report.js`** — `assertValidFix()` no longer builds a `JSON.stringify` message for every *valid* fix; one `RuleFixer` is shared per file report instead of one per fix/suggestion; drops a defensive descriptor clone (the descriptor is consumed synchronously and never retained or mutated).

- **`lib/linter/interpolate.js`** — skips the placeholder regex when a message contains no `{{`, and reuses one shared regex instead of constructing one per call.

- **`lib/linter/apply-disable-directives.js`** — fast path when a file has no disable directives (previously iterated all problems regardless).

- **`lib/shared/text-table.js`** — computes each cell's display width once instead of twice; uses `" ".repeat()` instead of `Array(n).join(" ")`.

- **`lib/cli-engine/formatters/stylish.js`** — skips `stripVTControlCharacters()` for strings containing neither the 7-bit ESC nor the 8-bit CSI introducer, and strips a trailing message period without a regex.

`node Makefile.js mocha` passes in full (38,574 passing, coverage thresholds met) and `node Makefile.js fuzz` reports no issues.

#### Is there anything you'd like reviewers to focus on?

- **`createFastMatcher()` in `lib/linter/esquery.js`** duplicates a slice of `esquery`'s matching semantics rather than delegating to it. A mismatch here would be a silent correctness bug rather than a crash, so it deserves a careful read. It bails out to the general path for anything other than the exact `Parent > .field` / `Parent > *.field` shape.

- **Error attribution via a symbol tag** instead of a wrapper closure (`linter.js` / `source-code-visitor.js`). The tag lives on the function object, so the fast path is skipped when a function is already tagged for a different rule. Worth confirming this covers plugin patterns you know of that aren't in the test suite.

- **Scope correction since the first push.** The original version of this PR also converted `lib/api.js` to lazy getters, which reported a much larger Loading win. That broke Node's CJS→ESM named-export interop (`cjs-module-lexer` cannot statically detect getter-based exports), so `import { RuleTester } from "eslint"` threw at runtime — caught by the ecosystem plugin tests and `attw`. It has been reverted; the Loading improvement now comes from the `ajv` change instead. The measurements above were re-taken after that revert.

## Summary by CodeRabbit

- **Performance**
  - Improved linting speed and reduced memory use across configuration loading, rule validation, source traversal, selector matching, formatting, and text rendering.
  - Improved handling of files without directives or interpolation placeholders.
  - Reduced startup overhead through deferred validation setup.
  - Optimized traversal of large syntax trees and processing of common selectors.

- **Bug Fixes**
  - Improved attribution of listener errors to the rule that generated them.
  - Preserved formatter alignment when messages contain terminal control characters.
  - Maintained correct report formatting and fix behavior while optimizing processing.

### CODE REVIEW

**Comment by tomasdev** [comment]
lib/api.js is mentioned in the description but it's not present in the changed files

File: lib/cli-engine/formatters/stylish.js [hunk 1]
```diff
@@ -28,6 +28,18 @@ function getStyleText(color) {
...
 }
 
+/**
+ * Computes the visible width of a string, avoiding the cost of stripping
+ * VT control characters when the string doesn't contain any.
+ * @param {string} str The string to measure.
+ * @returns {number} The number of visible characters.
+ */
+function stringLength(str) {
+	return str.includes("\u001b")
```
Comment (igouss) [inline lib/cli-engine/formatters/stylish.js]: Node's latest https://github.com/nodejs/node/blob/main/lib/internal/util/inspect.js#L3043 checks \u009B as well as \u001B

Comment (nzakas) [reply lib/cli-engine/formatters/stylish.js]: Good catch — confirmed and fixed in 89fdac26d.

Verified against Node v22.19.0: `util.stripVTControlCharacters()` strips CSI-introduced sequences as well, and a CSI-only string does not contain ESC — so the fast path returned the raw length (10 instead of 3 for a `CSI 31m red CSI 0m` string) and misaligned the table columns. The check now tests both introducers.

Added a regression test asserting rule-ID column alignment across rows when a message contains an 8-bit CSI introducer; it fails without the fix.

File: lib/cli-engine/formatters/stylish.js [hunk 2]
```diff
@@ -28,6 +28,18 @@ function getStyleText(color) {
...
+		? util.stripVTControlCharacters(str).length
+		: str.length;
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/cli-engine/formatters/stylish.js]: Fixed in 89fdac26d — same issue @igouss raised independently. Both the 7-bit ESC and 8-bit CSI introducers are now checked, with a regression test covering column alignment for a CSI-only message.

File: lib/linter/linter.js [hunk 3]
```diff
@@ -672,10 +672,30 @@ function runRules(
...
+			if (
+				!timing.enabled &&
+				!stats &&
+				typeof rawListener === "function" &&
+				Object.isExtensible(rawListener)
+			) {
+				rawListener[RULE_ID] = ruleId;
+				visitor.add(selector, rawListener);
+				return;
+			}
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/linter/linter.js:706]: Confirmed as a real bug and fixed in 89fdac26d.

I reproduced the misattribution before fixing it: with two rules whose `create()` returns the same function object, the second registration overwrote the tag, so an error thrown while running the *first* rule's listener was reported as the second rule.

The fast path now only applies when the function is untagged or already tagged for the same rule; otherwise the registration falls back to the per-registration wrapper, so each keeps its own rule ID.

Added a regression test (two rules sharing one listener function object) that fails without the guard.

File: lib/linter/source-code-visitor.js [hunk 4]
```diff
@@ -11,6 +11,27 @@
...
+ * Attributes an error thrown from a listener function to the rule that
+ * registered the listener, if known.
+ * @param {any} error The thrown value.
+ * @param {Function} func The listener function that threw the error.
+ * @returns {void}
+ */
+function attributeError(error, func) {
+	if (func[RULE_ID] !== void 0 && error.ruleId === void 0) {
+		error.ruleId = func[RULE_ID];
+	}
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/linter/source-code-visitor.js]: Fixed in 89fdac26d. You are right that this was an unintended behavior change — the wrapper it replaced always overwrote `ruleId` with the rule whose listener threw, so preserving a pre-existing value was a regression. `attributeError()` now assigns unconditionally, with a regression test covering a rule that throws an error already carrying a different `ruleId`.

File: lib/config/config.js [hunk 5]
```diff
@@ -12,8 +12,23 @@
...
+let ajv;
+
+/**
+ * Gets the lazily-constructed shared ajv instance.
+ * @returns {import("ajv").Ajv} The ajv instance.
+ */
+function getAjv() {
+	if (!ajv) {
+		ajv = require("../shared/ajv")();
+	}
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/config.js]: You're right that the description was stale, and thanks for catching it — the PR body has now been rewritten.

For the record on what happened: `lib/api.js` *was* originally converted to lazy getters, but that broke Node's CJS-to-ESM named-export interop — `cjs-module-lexer` can't statically detect getter-based exports, so `import { RuleTester } from "eslint"` threw a `SyntaxError` at runtime. That was caught by the ecosystem plugin tests (`@eslint/css`, `@eslint/json`, `@eslint/markdown`, `eslint-plugin-unicorn`) and the `attw` check, none of which are exercised by this repo's CJS-only unit tests. It was reverted in 897ccc7b9; breaking every ESM consumer isn't worth a startup win.

The Loading improvement now comes from making `ajv` lazy instead. Note this required changes in **two** places to have any effect — `lib/config/config.js` alone did nothing, because `lib/api.js` eagerly requires `RuleTester`, which independently constructed `ajv` at module scope. With both lazy (89fdac26d), the `require("eslint")` graph drops from 177 modules to 131.

The benchmark numbers in the description have been re-measured after the revert, using interleaved `hyperfine` runs against the baseline commit so machine drift can't be misattributed to the change: Loading 1.27×, Single File 1.12×, Multi Files 1.28×, aggregate −19.1%.

File: lib/linter/source-code-visitor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (igouss) [reply lib/linter/source-code-visitor.js]: unintended behavior change that did not trigger a test failure deserves a test.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/linter/source-code-visitor.js]: Fair point, and you're right about the root cause: `callSync()` had **no** direct coverage of error propagation at all. The attribution contract lives in this module, but the only tests touching it went through `Linter`, so changing the semantics here didn't fail anything.

Added unit tests in `tests/lib/linter/source-code-visitor.js` (b543b3098) covering the contract directly:

- an error thrown from an **untagged** function propagates with no `ruleId` added
- an error thrown from a **tagged** function is attributed to that rule
- an **existing `ruleId` is overwritten** by the rule that actually threw ← the regression
- attribution works on both the single-argument fast path and the multi-argument path (these are separate branches in `callSync()`)
- functions registered after a thrower are not called

I verified the overwrite tests fail against the previous preserve-if-unset behavior — they report `test/other-rule` where `test/some-rule` is expected — so they'd have caught the change that prompted your comment.

(Comment by Claude)

File: lib/cli-engine/formatters/stylish.js [hunk 6]
```diff
@@ -28,6 +28,22 @@ function getStyleText(color) {
...
+ * removes, so both must be checked before taking the fast path.
+ * @param {string} str The string to measure.
+ * @returns {number} The number of visible characters.
+ */
+function stringLength(str) {
+	return str.includes("\u001b") || str.includes("\u009b")
+		? util.stripVTControlCharacters(str).length
+		: str.length;
+}
+
```
Comment (fasttime) [inline lib/cli-engine/formatters/stylish.js:46]: I'm not sure if this is useful since `util.stripVTControlCharacters` already does the same optimization. See https://github.com/nodejs/node/blob/v26.7.0/lib/internal/util/inspect.js#L3046-L3051.

File: lib/config/config.js [hunk 7]
```diff
@@ -12,8 +12,23 @@
...
+	return ajv;
+}
```
Comment (fasttime) [inline lib/config/config.js]: This new code is now sitting between the `require` lines at the top of the file. Can we move it under the relevant sections "Private Members" and "Helpers" instead?

File: lib/linter/linter.js [hunk 8]
```diff
@@ -672,10 +672,38 @@ function runRules(
...
+			 * used when that object isn't already tagged for a different rule.
+			 * Rules that return the same function object (for example, a
+			 * shared visitor from a helper) fall back to the wrapper below so
+			 * that each registration keeps its own rule ID.
+			 */
+			if (
+				!timing.enabled &&
+				!stats &&
+				typeof rawListener === "function" &&
+				Object.isExtensible(rawListener) &&
```
Comment (fasttime) [inline lib/linter/linter.js]: Can we add unit tests where either of these conditions is false? Alternatively, if the conditions are always true, can we remove them instead?

File: lib/linter/source-code-visitor.js [hunk 9]
```diff
@@ -11,6 +11,33 @@
...
+ * @param {any} error The thrown value.
+ * @param {Function} func The listener function that threw the error.
+ * @returns {void}
+ */
+function attributeError(error, func) {
+	const ruleId = func[RULE_ID];
+
+	if (ruleId !== void 0) {
+		error.ruleId = ruleId;
+	}
```
Comment (fasttime) [inline lib/linter/source-code-visitor.js:80]: This will add a property to an object we don't control, where the consumer may not expect it, so I don't think it's a good idea. Would a `WeakMap` with `error` keys and `ruleId` values be a viable alternative?

One more thing: if `error` is a non-undefined primitive (for example a string), this assignment will throw a strict mode `TypeError`. That's probably unintended since the calling function wants to throw the `error` value passed in the first argument, and not a new error:

https://github.com/eslint/eslint/blob/89fdac26d75f436441e2c88c29c9b8b8d6aa61f1/lib/linter/source-code-visitor.js#L118-L119

Can you double-check the behavior?

Comment (nzakas) [reply lib/linter/source-code-visitor.js:80]: We were actually already doing this in the `ruleListener` closure, so this just moves the behavior.
However, a `WeakMap` probably works better in both cases. Good idea.

Comment (nzakas) [reply lib/linter/source-code-visitor.js:80]: Agreed on both counts — switched to a `WeakMap` in 71c77d27b.

**On the property:** you're right that writing to an object we don't own is the wrong trade for this. The mapping is now module-scoped (`WeakMap<Function, string>`), so nothing is added to the rule's function. Worth noting the hot path is unaffected either way — the mapping is only consulted inside the `catch`, never during normal listener invocation.

A nice side effect: `WeakMap` can key a frozen object, so the `Object.isExtensible()` guard was no longer needed and has been removed. Frozen listeners now take the fast path instead of falling back to a wrapper.

**On the primitive:** confirmed, and thanks for catching it. Before the fix:

```
rule throws "boom-as-string"
  -> TypeError: Cannot create property 'ruleId' on string 'boom-as-string'
```

The assignment is now guarded so non-object values are left alone.

**One caveat, because the fix is narrower than it first appears.** Guarding this assignment alone does *not* make primitives propagate end-to-end — it just moves where they're masked:

```
rule throws "boom-as-string"
  -> TypeError: Cannot create property 'currentNode' on string 'boom-as-string'
```

There are three places that decorate a thrown value, and all three have this issue:

1. this module (fixed here)
2. `SourceCodeTraverser#traverseSync` — `err.currentNode = step.target`
3. `Linter#_verifyWithFlatConfigArray` — `err.message += ...`

(2) and (3) are pre-existing and unrelated to the perf work, so I left them alone rather than expanding this PR's scope — and I removed the end-to-end test I'd written, since it would have asserted a fix that isn't actually complete. Happy to fix all three here or open a separate issue, whichever you prefer.

File: lib/cli-engine/formatters/stylish.js (hunk 6 above)
Comment (nzakas) [reply lib/cli-engine/formatters/stylish.js:46]: Good question — I checked, and the optimization landed in Node more recently than the versions ESLint supports.

The v26 source you linked does have it, but on Node v22.19.0 there's no short-circuit at all:

```
$ node -p "require('node:util').stripVTControlCharacters.toString()"
function stripVTControlCharacters(str) {
  validateString(str, 'str');

  return RegExpPrototypeSymbolReplace(ansi, str, '');
}
```

The regex runs unconditionally. Measured on v22.19.0, 300k calls with a plain (no-VT) string:

| | time |
| --- | --- |
| `stripVTControlCharacters` directly | 21.5 ms |
| guarded by `includes()` | 5.2 ms |
| `.length` (floor) | 1.4 ms |

So it's ~4× on this microbenchmark for supported versions. In the Single File benchmark that's roughly 312k calls (52k messages × 6 cells), worth on the order of ~15 ms out of ~1.9 s — real but small, under 1%.

On Node 24+/26, where the check exists upstream, this becomes a redundant (but very cheap) extra `includes()`. So the value decays as the supported range moves forward.

I've left it in since ESLint still supports `^20.19.0 || ^22.13.0`, but I don't feel strongly — it's the weakest optimization in the PR and it already caused one bug (the CSI case above). Happy to drop it for simplicity if you'd prefer; the `stripVTControlCharacters` call alone is correct on every version, and the CSI regression test stays valid either way.

File: lib/linter/linter.js (hunk 8 above)
Comment (nzakas) [reply lib/linter/linter.js]: Both, as it turns out — one condition was removable and the other was reachable and now has a test (71c77d27b).

**`Object.isExtensible(rawListener)` — removed.** It was only there because the rule ID was stored as a property on the function, which fails on a frozen function. Switching to a `WeakMap` (per your other comment) makes it unnecessary, since a `WeakMap` can key a frozen object. Frozen listeners now take the fast path instead of falling back to a wrapper, and there's a test asserting a frozen listener's errors are still attributed correctly.

**`typeof rawListener === "function"` — kept, and now tested.** It is reachable: a rule can return a non-function value for a selector, e.g. `create: () => ({ Program: null })`. That currently fails at call time with `TypeError: ruleListener is not a function` (same as before this PR), and the condition is what routes it down the wrapper path. It's also required for correctness now — `WeakMap.prototype.set` throws on a primitive key, so tagging a non-object would turn a clear "not a function" error into a confusing `Invalid value used as weak map key`.

Added tests for both cases in `tests/lib/linter/linter.js`. I verified the frozen-listener test passes with `Object.isExtensible()` either present or absent, which is what makes removing it safe rather than just untested.

File: lib/config/config.js (hunk 7 above)
Comment (nzakas) [reply lib/config/config.js]: Done in 71c77d27b — moved the `ajv` variable under "Private Members" and `getAjv()` under "Helpers". Applied the same treatment in `lib/rule-tester/rule-tester.js`, where the lazy `ajv` code had the same problem.

File: lib/config/config.js [hunk 10]
```diff
@@ -596,7 +623,14 @@ class Config {
 	 * @returns {RuleDefinition|undefined} The rule definition from the plugin, or `undefined` if the rule is not found.
 	 */
 	getRuleDefinition(ruleId) {
-		return getRuleFromConfig(ruleId, this);
```
Comment (mdjermanovic) [inline lib/config/config.js:599]: There are two more places that call `getRuleFromConfig(ruleId, this)`:

https://github.com/eslint/eslint/blob/9aa38732177935bd1d7f1493732c0b67666be28a/lib/config/config.js#L618

https://github.com/eslint/eslint/blob/9aa38732177935bd1d7f1493732c0b67666be28a/lib/config/config.js#L688


Perhaps we could update them to call `this.getRuleDefinition(ruleId)` in order to use the same cache?

Comment (nzakas) [reply lib/config/config.js:599]: Good catch — done in 1f66bcf75. Both now call `this.getRuleDefinition(ruleId)`, which leaves `getRuleFromConfig()` with a single caller, so the cache is the only lookup path.

The effect is larger than it looks, because both of those run over *every* configured rule, so each rule was being looked up three times. Counting actual `getRuleFromConfig()` executions during a lint with all core rules enabled:

| | executions |
| --- | --- |
| before | 876 |
| after | 292 |

Exactly 3× — one per rule instead of three, each of which repeated the rule ID parsing and the plugin proxy traps.

Worth noting for anyone reading later: this is per `Config` construction rather than per file, so it's startup cost rather than per-file cost.

I checked the two things that could have made this unsafe:

- **Staleness** — `this.plugins` is assigned once in the constructor and never reassigned, so a cached definition can't go stale. The lazy rule proxy in `default-config.js` also returns a stable object once resolved.
- **Initialization order** — `#normalizeRulesConfig()` runs from the constructor, but `#ruleDefinitions` is a class field, so it's initialized before the constructor body executes.

One incidental benefit: the cache is a `Map`, so the `__proto__` rule ID edge case (there's an existing test for it) is handled as an ordinary key rather than hitting `Object.prototype`.

Full suite passes (38,584), coverage thresholds met.

File: lib/cli-engine/formatters/stylish.js (hunk 6 above)
Comment (fasttime) [reply lib/cli-engine/formatters/stylish.js:46]: Ah, I see, thanks for checking this. In that case it's fine to leave the optimization as is, I think. Maybe we could add a comment to say that the checks can be removed once Node.js<24.15.0 is dropped.

File: lib/linter/source-code-visitor.js [hunk 11]
```diff
@@ -70,12 +137,49 @@ class SourceCodeVisitor {
...
+
+		for (let i = 0; i < funcs.length; i++) {
+			const func = funcs[i];
+
+			try {
+				func(...args);
+			} catch (err) {
+				attributeError(err, func);
+				throw err;
+			}
```
Comment (fasttime) [inline lib/linter/source-code-visitor.js:175]: If we are going to keep separate implementations for the single-argument and multi-argument cases for performance reasons, I think we might as well go further and add a separate `callSyncWithSingleArg` method alongside `callSync` rather than branching once the method is called.

This seems possible because the number of arguments is statically known at each call site: there are two invocations where `callSync` is called with exactly one argument after `source`:

https://github.com/eslint/eslint/blob/1f66bcf7506e534b8e6f5cdb4948bf467dc2ccc2/lib/linter/source-code-traverser.js#L330-L333

In the other places, it's always called with more than one argument:

https://github.com/eslint/eslint/blob/1f66bcf7506e534b8e6f5cdb4948bf467dc2ccc2/lib/linter/source-code-traverser.js#L335

https://github.com/eslint/eslint/blob/1f66bcf7506e534b8e6f5cdb4948bf467dc2ccc2/lib/linter/source-code-traverser.js#L350

File: lib/linter/source-code-visitor.js (hunk 9 above)
Comment (fasttime) [reply lib/linter/source-code-visitor.js:80]: Oh, if there are more places where an error can be thrown by trying to set a property on a primitive, then I'd say it's better to fix them in a separate PR.

File: lib/linter/esquery.js [hunk 12]
```diff
@@ -230,6 +253,95 @@ function analyzeParsedSelector(parsedSelector) {
...
+	if (right.type === "field") {
+		field = right;
+	} else if (right.type === "compound" && right.selectors.length === 2) {
+		const [first, second] = right.selectors;
+
+		if (first.type === "wildcard" && second.type === "field") {
+			field = second;
+		} else if (first.type === "field" && second.type === "wildcard") {
+			field = first;
+		}
```
Comment (mdjermanovic) [inline lib/linter/esquery.js:311]: Can this case happen and how would a selector that falls into this branch look like?

File: lib/linter/esquery.js [hunk 13]
```diff
@@ -230,6 +253,95 @@ function analyzeParsedSelector(parsedSelector) {
...
+ * @returns {Function|null} A matcher function, or `null` if the selector doesn't have the expected shape.
+ */
+function createFastMatcher(root) {
+	if (root.type !== "child") {
+		return null;
+	}
+
+	const { left, right } = root;
+
+	if (left.type !== "identifier" || left.subject || right.subject) {
```
Comment (mdjermanovic) [inline lib/linter/esquery.js:296]: We don't support subject indicators, but since we're checking them here, why is `right.subject` problematic for this case? I think it's effectively a no-op.

File: lib/linter/esquery.js [hunk 14]
```diff
@@ -230,6 +253,95 @@ function analyzeParsedSelector(parsedSelector) {
...
+		const [first, second] = right.selectors;
+
+		if (first.type === "wildcard" && second.type === "field") {
+			field = second;
+		} else if (first.type === "field" && second.type === "wildcard") {
+			field = first;
+		}
+	}
+
+	if (!field || field.subject || field.name.includes(".")) {
```
Comment (mdjermanovic) [inline lib/linter/esquery.js:314]: Same as my previous comment, I think `field.subject` is effectively a no-op.

File: lib/linter/source-code-traverser.js [hunk 15]
```diff
@@ -275,39 +294,50 @@ class SourceCodeTraverser {
 		});
 
 		const currentAncestry = [];
+		const matchedSelectors = [];
+		const allSteps = steps ?? sourceCode.traverse();
+
+		// Iterating an array by index avoids iterator protocol overhead.
+		const stepsArray = Array.isArray(allSteps)
+			? allSteps
+			: Array.from(allSteps);
```
Comment (mdjermanovic) [inline lib/linter/source-code-traverser.js:303]: For non-arrays, would this use the iterator protocol via `Array.from()` and create an extra array, which would degrade performance overall compared to the previous version? If returning arrays from `sourceCode.traverse()` significantly improves performance, perhaps we should mention it as the recommendation in the docs, and also update our language plugins to return arrays.

File: lib/linter/source-code-visitor.js (hunk 11 above)
Comment (nzakas) [reply lib/linter/source-code-visitor.js:175]: I'd prefer not to do that because I envision `callSync` as something we may expose as part of the API in the future (along with an async `call()`). Branching on argument length is a cheap operation so I don't think splitting buys us much anyway.

File: lib/linter/source-code-traverser.js (hunk 15 above)
Comment (nzakas) [reply lib/linter/source-code-traverser.js]: Ooh good catch. Yes, we should not be converting iterables into arrays unnecessarily.

File: lib/linter/esquery.js (hunk 12 above)
Comment (nzakas) [reply lib/linter/esquery.js]: You're right, this can't happen. Fixed.

File: lib/linter/esquery.js (hunk 13 above)
Comment (nzakas) [reply lib/linter/esquery.js]: Right again, removed.

File: lib/linter/esquery.js (hunk 14 above)
Comment (nzakas) [reply lib/linter/esquery.js]: Yup, removed.

**Review by mdjermanovic — DISMISSED** [review]
LGTM, aside from the suggestion to add a comment that the `stringLength()` fast path can be removed once support for Node.js<24.15.0 is dropped (https://github.com/eslint/eslint/pull/21210#discussion_r3822158488), which I agree with.

I'm also not sure if the PR title best describes the changes, since most of them are not CLI-specific.

**Comment by nzakas** [comment]
Added the comment.

**Comment by mdjermanovic** [comment]
I'll close & reopen this to re-run CI with https://github.com/eslint/eslint/commit/92086c87e042413a3d0363d8fc3fbb11db98d06a.

**Review by mdjermanovic — APPROVED** [review]

**Review by fasttime — APPROVED** [review]
LGTM, thanks!
